internal/controller/postgrescluster/snapshots.go

```diff
@@ -16,53 +18,82 @@ import (
+//     timestamp.
+//  5. If there are no snapshots at all, we take a snapshot and put the backup job's completion
+//     timestamp on the snapshot annotation.
+//  6. If an earlier snapshot is found, we take a new snapshot, annotate it and delete the old
+//     snapshot.
+//  7. When the snapshot job completes, we delete the restore job.
 func (r *Reconciler) reconcileVolumeSnapshots(ctx context.Context,
-	postgrescluster *v1beta1.PostgresCluster, instances *observedInstances,
-	clusterVolumes []corev1.PersistentVolumeClaim) error {
+	postgrescluster *v1beta1.PostgresCluster, pvc *corev1.PersistentVolumeClaim) error {
 
 	// Get feature gate state
 	volumeSnapshotsFeatureEnabled := feature.Enabled(ctx, feature.VolumeSnapshots)
 
+	// TODO(dsessler7): If user has a snapshots section in their cluster spec, indicating
+	// that they would like to use snapshots, but the VolumeSnapshots feature gate is not
+	// turned on, should we emit an event or log something?
+
 	// Check if the Kube cluster has VolumeSnapshots installed. If VolumeSnapshots
-	// are not installed we need to return early. If user is attempting to use
+	// are not installed, we need to return early. If user is attempting to use
 	// VolumeSnapshots, return an error, otherwise return nil.
-	volumeSnapshotsExist, err := r.GroupVersionKindExists("snapshot.storage.k8s.io/v1", "VolumeSnapshot")
+	volumeSnapshotKindExists, err := r.GroupVersionKindExists("snapshot.storage.k8s.io/v1", "VolumeSnapshot")
 	if err != nil {
 		return err
 	}
-	if !*volumeSnapshotsExist {
+	if !*volumeSnapshotKindExists {
 		if postgrescluster.Spec.Backups.Snapshots != nil && volumeSnapshotsFeatureEnabled {
 			return errors.New("VolumeSnapshots are not installed/enabled in this Kubernetes cluster; cannot create snapshot.")
 		} else {
 			return nil
 		}
 	}
 
-	// Get all snapshots for this cluster
+	// Get all snapshots for the cluster.
 	snapshots, err := r.getSnapshotsForCluster(ctx, postgrescluster)
 	if err != nil {
 		return err
 	}
 
 	// If snapshots are disabled, delete any existing snapshots and return early.
 	if postgrescluster.Spec.Backups.Snapshots == nil || !volumeSnapshotsFeatureEnabled {
-		for i := range snapshots.Items {
-			if err == nil {
-				err = errors.WithStack(client.IgnoreNotFound(
-					r.deleteControlled(ctx, postgrescluster, &snapshots.Items[i])))
-			}
-		}
+		return r.deleteSnapshots(ctx, postgrescluster, snapshots)
+	}
+
+	// If we got here, then the snapshots are enabled (feature gate is enabled and the
+	// cluster has a Spec.Backups.Snapshots section defined).
 
+	// The specified VolumeSnapshotClass needs to exist.
+	volumeSnapshotClass := &volumesnapshotv1.VolumeSnapshotClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: postgrescluster.Spec.Backups.Snapshots.VolumeSnapshotClassName,
+		},
+	}
+	err = errors.WithStack(r.Client.Get(ctx, client.ObjectKeyFromObject(volumeSnapshotClass),
+		volumeSnapshotClass))
```

Comment:
Yes, this is why we need those permissions on `volumesnapshotclass`.

I believe it will fail/error if you attempt to use a `volumesnapshotclass` that doesn't exist (I will double check this).

Are you thinking that we don't bother checking that the class exists but just pass the error/create an event when the snapshot creation fails due to a non-existent `volumesnapshotclass` being specified?

Reply:
@dsessler7 What happens when user leaves `volumeSnapshotClass` blank/empty string?

Reply:
It errors and does not attempt to create a snapshot if you give an empty string for `volumeSnapshotClassName`: 
```
ERRO[2024-10-01T12:48:49-07:00] Reconciler error                              PostgresCluster=postgres-operator/hippo controller=postgrescluster controllerGroup=postgres-operator.crunchydata.com controllerKind=PostgresCluster error="admission webhook \"validation-webhook.snapshot.storage.k8s.io\" denied the request: Spec.VolumeSnapshotClassName must not be the empty string" file="internal/controller/postgrescluster/snapshots.go:145" func="postgrescluster.(*Reconciler).reconcileVolumeSnapshots" name=hippo namespace=postgres-operator reconcileID=642232f6-ca1f-40df-8e63-8d51ddc4bc40 version=
```
However, if you give a name for a non-existent `VolumeSnapshotClass` then it will create a VolumeSnapshot object, but the snapshot will have an error in its `Status`. We currently create a warning event when we detect an error in any of the existing VolumeSnapshots, but we don't log an error and we don't stop the operator creating more snapshots... Should this behavior change in some way? :thinking:

Reply:
> Should this behavior change in some way? 🤔

We can prevent the empty string with API validation; a `+kubebuilder:validation:MinLength=1` in here:

https://github.com/CrunchyData/postgres-operator/blob/66174ec98044d395bef1165a69f2fc11139c0a5f/pkg/apis/postgres-operator.crunchydata.com/v1beta1/postgrescluster_types.go#L693-L698